### PR TITLE
solution was added

### PR DESCRIPTION
**Pull request by Maksym845**
I have no idea why tests fail when I rewrite my server on functions, instead classes.

### CODE REVIEW

**Review by Oleksandra-Shevchenko — APPROVED** [review]
GJ 💯

File: src/services/expense.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+class Expense {
```
Comment (OleksiiBlezniuk) [inline src/services/expense.js:3]: From `Expense` name I expect that I this is an object of Expense interface or it's class/constructor

Comment (OleksiiBlezniuk) [inline src/services/expense.js:3]:
```suggestion
class ExpenseService {
```

File: src/services/users.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+class Users {
```
Comment (OleksiiBlezniuk) [inline src/services/users.js:3]:
```suggestion
class UsersService {
```

File: src/createServer.js [hunk 3]
```diff
@@ -1,11 +1,144 @@
...
+
+    if (!user) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    res.statusCode = 200;
+    res.send(user);
+  });
```
Comment (OleksiiBlezniuk) [inline src/createServer.js:40]: Let's create routers for every entity and controllers for actions 🙂

File: src/createServer.js [hunk 4]
```diff
@@ -1,11 +1,144 @@
...
+
+  app.delete('/users/:id', express.json(), (req, res) => {
+    const user = users.deleteUser(req.params.id);
+
+    if (user) {
+      res.statusCode = 204;
```
Comment (OleksiiBlezniuk) [inline src/createServer.js:46]: why not `sendStatus`?

File: src/createServer.js [hunk 5]
```diff
@@ -1,11 +1,144 @@
...
+    }
+
+    res.sendStatus(404);
+  });
+
+  app.patch('/users/:id', express.json(), (req, res) => {
+    const user = users.updateUser(req.params.id, req.body);
+
+    if (user) {
+      res.statusCode = 200;
```
Comment (OleksiiBlezniuk) [inline src/createServer.js:59]: this status will be sent by default

File: src/createServer.js [hunk 6]
```diff
@@ -1,11 +1,144 @@
...
+  });
+
+  app.post('/expenses', express.json(), (req, res) => {
+    const { title } = req.body;
+
+    if (!title) {
+      res.sendStatus(400);
+
+      return;
+    }
```
Comment (OleksiiBlezniuk) [inline src/createServer.js:88]: What about other properties?

File: src/createServer.js [hunk 7]
```diff
@@ -1,11 +1,144 @@
...
+    const expense = expenses.updateExpense(req.params.id, req.body);
+
+    if (expense) {
+      res.statusCode = 200;
+      res.send(expense);
+
+      return;
+    }
+
+    res.sendStatus(404);
```
Comment (OleksiiBlezniuk) [inline src/createServer.js:125]: Better to validate data before updating data in storage, and not vice versa 🙂

File: src/services/expense.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  constructor() {
+    this.expenses = [];
+  }
+
+  getExpenses(query) {
+    if (query.userId && query.category) {
```
Comment (OleksiiBlezniuk) [inline src/services/expense.js:9]: destructure query object

File: src/services/expense.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      return this.expenses.filter(expense => expense.userId === +query.userId
```
Comment (OleksiiBlezniuk) [inline src/services/expense.js:10]: move conditions inside filter instead

File: src/services/expense.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      this.expenses.push(newExpense);
+
+      return newExpense;
+    }
+
+    return false;
+  }
+
+  getExpenseById(expenseId) {
+    return this.expenses.find(expense => expense.id === +expenseId);
```
Comment (OleksiiBlezniuk) [inline src/services/expense.js:52]:
```suggestion
    return this.expenses.find(expense => expense.id === Number(expenseId));
```

File: src/services/expense.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    const foundExpense = this.expenses
+      .find(expense => expense.id === +expenseId);
+
+    if (foundExpense) {
+      Object.assign(foundExpense, newData);
+
+      return foundExpense;
+    }
+
+    return false;
```
Comment (OleksiiBlezniuk) [inline src/services/expense.js:65]: It's strange that your function return object or boolean, in that case more clear to return null if action is not performed

File: src/services/users.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    let id = 0;
+
+    if (this.users.length !== 0) {
+      id = this.users[this.users.length - 1].id + 1;
+    }
+
+    const newUser = {
+      id,
+      name: newUserName,
+    };
```
Comment (OleksiiBlezniuk) [inline src/services/users.js:22]: Can id be `0`?

**Review by OleksiiBlezniuk — CHANGES_REQUESTED** [review]

**Review by OleksiiBlezniuk — APPROVED** [review]

**Pull request closed without merging**